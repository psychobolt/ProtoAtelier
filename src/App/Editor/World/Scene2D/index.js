export * from './Scene2D.component';
export { default } from './Scene2D.component';
