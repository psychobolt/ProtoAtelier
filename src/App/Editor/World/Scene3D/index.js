export { default } from './Scene3D.component';
