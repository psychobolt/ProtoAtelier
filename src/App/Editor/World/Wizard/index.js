export { default } from './Wizard.component';
