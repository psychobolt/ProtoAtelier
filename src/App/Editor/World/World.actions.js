import Hashids from 'hashids';

export const Actions = {
  UPDATE_GRAVITY: 'updateGravity',
  SET_PIXELS_PER_METER: 'setPixelsPerMeter',
  ADD_BODY: 'addBody',
  REMOVE_BODIES: 'removeBodies',
  UPDATE_BODIES: 'updateBodies',
  ADD_BODY_LINKS: 'addBodyLinks',
  ADD_FIXTURES: 'addFixtures',
  REMOVE_FIXTURES: 'removeFixtures',
  UPDATE_FIXTURES: 'updateFixtures',
  SELECT_BODY: 'selectBody',
  ADD_JOINT: 'addJoint',
  REMOVE_JOINTS: 'removeJoints',
  UPDATE_JOINTS: 'updateJoints',
  ADD_ANCHOR: 'addAnchor',
  REMOVE_ANCHORS: 'removeAnchors',
  UPDATE_ANCHORS: 'updateAnchors',
  TOGGLE_PAUSE: 'togglePause',
  TOGGLE_TRACKING: 'toggleTracking',
  UPDATE_MOUSE_FORCE: 'updateMouseForce',
  UPDATE_SPEED: 'updateSpeed',
  RESET: 'reset',
};

export const updateGravity = gravity => ({ type: Actions.UPDATE_GRAVITY, payload: gravity });

export const setPixelsPerMeter = value => ({
  type: Actions.SET_PIXELS_PER_METER, payload: { value },
});

export const addBody = skipHistory => ({
  type: Actions.ADD_BODY,
  payload: {
    id: new Hashids('body').encode(new Date().getTime()),
    fixtureIds: [],
    type: 'static',
    linearVelocity: { x: 0, y: 0 },
    linearDamping: 0,
    angularDamping: 0,
    bullet: false,
  },
  meta: { skipHistory },
});

export const removeBodies = ids => ({ type: Actions.REMOVE_BODIES, payload: { ids } });

export const updateBodies = bodies => ({ type: Actions.UPDATE_BODIES, payload: bodies });

export const addFixtures = (bodyId, fixtures, skipHistory) => ({
  type: Actions.ADD_FIXTURES,
  payload: {
    bodyId,
    fixtures: fixtures.map((fixture, index) => ({
      id: new Hashids('fixture').encode(index, new Date().getTime()),
      bodyId,
      density: 0,
      friction: 0,
      restitution: 0,
      ...fixture,
    })),
  },
  meta: { skipHistory },
});

export const removeFixtures = (ids, skipHistory) => ({
  type: Actions.REMOVE_FIXTURES, payload: { ids }, meta: { skipHistory },
});

export const updateFixtures = fixtures => ({ type: Actions.UPDATE_FIXTURES, payload: fixtures });

export const selectBody = id => ({
  type: Actions.SELECT_BODY, payload: { id }, meta: { skipHistory: true },
});

export const addJoint = options => ({
  type: Actions.ADD_JOINT,
  payload: {
    id: new Hashids('joint').encode(new Date().getTime()),
    anchors: {},
    enableMotor: false,
    maxMotorTorque: 0,
    motorSpeed: 0,
    ...options,
  },
});

export const removeJoints = ids => ({ type: Actions.REMOVE_JOINTS, payload: { ids } });

export const updateJoints = joints => ({ type: Actions.UPDATE_JOINTS, payload: joints });

export const addAnchor = (jointId, anchor) => ({
  type: Actions.ADD_ANCHOR,
  payload: {
    id: new Hashids('anchor').encode(new Date().getTime()),
    jointId,
    ...anchor,
  },
});

export const removeAnchors = (jointId, anchorIds) => ({
  type: Actions.REMOVE_ANCHORS, payload: { jointId, anchorIds },
});

export const updateAnchors = (jointId, anchors) => ({
  type: Actions.UPDATE_ANCHORS, payload: { jointId, anchors },
});

export const togglePause = () => ({ type: Actions.TOGGLE_PAUSE });

export const toggleTracking = () => ({ type: Actions.TOGGLE_TRACKING });

export const updateMouseForce = force => ({ type: Actions.UPDATE_MOUSE_FORCE, payload: force });

export const updateSpeed = speed => ({ type: Actions.UPDATE_SPEED, payload: speed });

export const reset = () => ({ type: Actions.RESET, payload: { id: new Hashids('world').encode(new Date().getTime()) } });
