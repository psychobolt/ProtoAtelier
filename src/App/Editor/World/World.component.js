// @flow
import React from 'react';
import { type Location } from 'react-router-dom';
import {
  SceneModule,
  DefineModule,
  PerspectiveCamera,
  RenderingModule,
  OrbitControlsModule,
  ResizeModule,
} from 'whs/build/whs';
import { App } from 'react-whs';
import * as THREE from 'three';

import Fragment from './World.fragment';
import styles from './World.styles';

const containerId = 'world-container';

type Props = {
  style: {
    container: {}
  },
  location: Location
};

type State = {
  mounted: boolean
};

export default class World extends React.Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      mounted: false,
    };
  }

  componentDidMount() {
    this.onMount(() => this.setState({ mounted: true }));
  }

  onMount = (callback: () => mixed) => {
    callback();
  }

  container: ?HTMLDivElement;

  render() {
    const { style, location } = this.props;
    const { mounted } = this.state;
    const container = this.container;
    const box: ?ClientRect = container && container.getBoundingClientRect();
    return (
      <div id={containerId} style={style.container} ref={ref => { this.container = ref; }}>
        {mounted && box && <App
          modules={[
            new SceneModule(),
            new DefineModule('camera', new PerspectiveCamera({
              aspect: box.width / box.height,
              position: new THREE.Vector3(0, 10, 50),
            })),
            new RenderingModule({
              bgColor: 0x162129,
              renderer: {
                antialias: true,
                shadowmap: {
                  type: THREE.PCFSoftShadowMap,
                },
              },
              width: box.width,
              height: box.height,
            }, { shadow: true }),
            new OrbitControlsModule(),
          ]}
          parentStyle={styles.parent}
          passAppToView={({ native }) => {
            native.manager.set('container', container);
            native.applyModule(new ResizeModule());
          }}
        >
          {Fragment({ location })}
        </App>}
      </div>
    );
  }
}
