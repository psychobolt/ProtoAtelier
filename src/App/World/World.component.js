// @flow
import React from 'react';
import { type Location } from 'react-router-dom';
import {
  Box,
  Sphere,
  Plane,
  PointLight,
  AmbientLight,
} from 'react-whs';
import * as THREE from 'three';

import withResizableContainer from 'Framework/ReactResizableContainer';
import Scene from './Scene';
import styles from './World.styles';

type Props = {
  containerEl: HTMLDivElement,
  containerWidth: number,
  containerHeight: number,
  location: Location
};

export const World =
({ containerEl, containerWidth = 680, containerHeight = 420, location }: Props) => (
  <Scene
    container={containerEl}
    width={containerWidth}
    height={containerHeight}
    style={styles.container}
  >
    {location && location.pathname === '/sphere' ?
      <Sphere
        key="1"
        geometry={{
          radius: 5,
          widthSegments: 32,
          heightSegments: 32,
        }}
        material={new THREE.MeshPhongMaterial({ color: 0xF2F2F2 })}
        position={{
          y: 5,
        }}
      /> :
      <Box
        key="2"
        geometry={{
          width: 10,
          height: 10,
          depth: 10,
        }}
        material={new THREE.MeshPhongMaterial({ color: 0xF2F2F2 })}
        position={{
          y: 5,
        }}
      />}
    <Plane
      key="3"
      geometry={{
        width: 100,
        height: 100,
      }}
      material={new THREE.MeshPhongMaterial({ color: 0x447F8B })}
      rotation={{
        x: -Math.PI / 2,
      }}
    />
    <PointLight
      key="4"
      intensity={0.5}
      distance={100}
      shadow={{
        fow: 90,
      }}
      position={new THREE.Vector3(0, 10, 10)}
    />
    <AmbientLight
      key="5"
      intensity={0.4}
    />
  </Scene>
);

export default withResizableContainer(World);
