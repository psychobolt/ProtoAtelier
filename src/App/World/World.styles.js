export default {
  parent: {
    width: '100%',
    height: '100%',
  },
};
