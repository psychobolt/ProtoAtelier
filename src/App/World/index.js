export { default } from './World.component';
