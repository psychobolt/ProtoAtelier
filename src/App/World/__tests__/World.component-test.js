import React from 'react';
import { shallow, mount } from 'enzyme';

import { World } from '../World.component';

describe('component <World />', () => {
  it('should render correctly with Box by default', () => {
    const wrapper = mount(<World />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly with Sphere', () => {
    const props = {
      location: {
        pathname: '/sphere',
      },
    };
    const wrapper = shallow(<World {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
