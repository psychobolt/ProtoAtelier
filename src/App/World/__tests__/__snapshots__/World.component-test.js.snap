// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`component <World /> should render correctly with Box by default 1`] = `
<Scene
  height={420}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  width={680}
>
  <r
    geometry={
      Object {
        "depth": 10,
        "height": 10,
        "width": 10,
      }
    }
    key="2"
    material={
      Object {
        "color": 15921906,
        "depthFunc": 3,
        "depthTest": true,
        "depthWrite": true,
        "emissive": 0,
        "metadata": Object {
          "generator": "Material.toJSON",
          "type": "Material",
          "version": 4.5,
        },
        "shininess": 30,
        "specular": 1118481,
        "type": "MeshPhongMaterial",
        "uuid": 0,
      }
    }
    position={
      Object {
        "y": 5,
      }
    }
  />
  <r
    geometry={
      Object {
        "height": 100,
        "width": 100,
      }
    }
    key="3"
    material={
      Object {
        "color": 4489099,
        "depthFunc": 3,
        "depthTest": true,
        "depthWrite": true,
        "emissive": 0,
        "metadata": Object {
          "generator": "Material.toJSON",
          "type": "Material",
          "version": 4.5,
        },
        "shininess": 30,
        "specular": 1118481,
        "type": "MeshPhongMaterial",
        "uuid": 0,
      }
    }
    rotation={
      Object {
        "x": -1.5707963267948966,
      }
    }
  />
  <r
    distance={100}
    intensity={0.5}
    key="4"
    position={
      Vector3 {
        "x": 0,
        "y": 10,
        "z": 10,
      }
    }
    shadow={
      Object {
        "fow": 90,
      }
    }
  />
  <r
    intensity={0.4}
    key="5"
  />
</Scene>
`;

exports[`component <World /> should render correctly with Sphere 1`] = `
<Scene
  height={420}
  style={
    Object {
      "height": "100%",
      "width": "100%",
    }
  }
  width={680}
>
  <r
    geometry={
      Object {
        "heightSegments": 32,
        "radius": 5,
        "widthSegments": 32,
      }
    }
    key="1"
    material={
      Object {
        "color": 15921906,
        "depthFunc": 3,
        "depthTest": true,
        "depthWrite": true,
        "emissive": 0,
        "metadata": Object {
          "generator": "Material.toJSON",
          "type": "Material",
          "version": 4.5,
        },
        "shininess": 30,
        "specular": 1118481,
        "type": "MeshPhongMaterial",
        "uuid": 0,
      }
    }
    position={
      Object {
        "y": 5,
      }
    }
  />
  <r
    geometry={
      Object {
        "height": 100,
        "width": 100,
      }
    }
    key="3"
    material={
      Object {
        "color": 4489099,
        "depthFunc": 3,
        "depthTest": true,
        "depthWrite": true,
        "emissive": 0,
        "metadata": Object {
          "generator": "Material.toJSON",
          "type": "Material",
          "version": 4.5,
        },
        "shininess": 30,
        "specular": 1118481,
        "type": "MeshPhongMaterial",
        "uuid": 0,
      }
    }
    rotation={
      Object {
        "x": -1.5707963267948966,
      }
    }
  />
  <r
    distance={100}
    intensity={0.5}
    key="4"
    position={
      Vector3 {
        "x": 0,
        "y": 10,
        "z": 10,
      }
    }
    shadow={
      Object {
        "fow": 90,
      }
    }
  />
  <r
    intensity={0.4}
    key="5"
  />
</Scene>
`;
